### ISSUE TITLE
CVE-2017-0249 Elevation of Privilege issue

### ISSUE THREAD

**Issue by RaghuMasineni** [dotnet/corefx#41019]
we have developed our application on asp.net core 1.1 version with VS2015.

We are facing Elevation of Privilege issue for System.Net.Security 4.3.0 in my application while deploying build using Jenkins. So I have created a direct dependency for System.Net.Security by changing its version to 4.3.1 from 4.3.0 in project.json file as below,

  "netcoreapp1.1": {
      "dependencies": {
        "Microsoft.NETCore.App": {
          "type": "platform",
          "version": "1.1.0"
        },
        "System.Net.Security": "4.3.1"
      }
    }

But I am still facing Elevation of Privilege issue with new warning messages for runtime/win/unix 
paths of System.Net.Security.

What I see is in  project.lock.json is that version 4.3.1 is not getting reflected after adding direct dependency in project.json as seen below,

"System.Net.Security/4.3.1": {
        "type": "package",
        "dependencies": {
          "Microsoft.NETCore.Platforms": "1.1.0",
          "Microsoft.Win32.Primitives": "4.3.0",
          "System.Collections": "4.3.0",
          "System.Collections.Concurrent": "4.3.0",
          "System.Diagnostics.Tracing": "4.3.0",
          "System.Globalization": "4.3.0",
          "System.Globalization.Extensions": "4.3.0",
          "System.IO": "4.3.0",
          "System.Net.Primitives": "4.3.0",
          "System.Resources.ResourceManager": "4.3.0",
          "System.Runtime": "4.3.0",
          "System.Runtime.Extensions": "4.3.0",
          "System.Runtime.Handles": "4.3.0",
          "System.Runtime.InteropServices": "4.3.0",
          "System.Security.Claims": "4.3.0",
          "System.Security.Cryptography.Algorithms": "4.3.0",
          "System.Security.Cryptography.Encoding": "4.3.0",
          "System.Security.Cryptography.OpenSsl": "4.3.0",
          "System.Security.Cryptography.Primitives": "4.3.0",
          "System.Security.Cryptography.X509Certificates": "4.3.0",
          "System.Security.Principal": "4.3.0",
          "System.Text.Encoding": "4.3.0",
          "System.Threading": "4.3.0",
          "System.Threading.Tasks": "4.3.0",
          "System.Threading.ThreadPool": "4.3.0",
          "runtime.native.System": "4.3.0",
          "runtime.native.System.Net.Security": "4.3.0",
          "runtime.native.System.Security.Cryptography.OpenSsl": "4.3.0"
        },
        "compile": {
          "ref/netstandard1.3/System.Net.Security.dll": {}
        },
        "runtimeTargets": {
          "runtimes/unix/lib/netstandard1.6/System.Net.Security.dll": {
            "assetType": "runtime",
            "rid": "unix"
          },
          "runtimes/win/lib/netstandard1.3/System.Net.Security.dll": {
            "assetType": "runtime",
            "rid": "win"
          }
        }
      }

Please let me know if I need to make any additional changes apart of adding direct dependency.

**Comment by danmosemsft** [comment]
@joperezr for question above

**Comment by joperezr** [comment]
What do you mean it is not getting reflected? The very first line: `"System.Net.Security/4.3.1"` means that the package that was restored was 4.3.1. The rest of the 4.3.0 versions just mean that the 4.3.1 System.Net.Security package depends on some packages that are 4.3.0. This matches the information in NuGet.org: https://www.nuget.org/packages/System.Net.Security/4.3.1 (open dependencies on that link to see)

Apart from that, is there a reason why you are still using project.json? We would strongly recommend you migrating to the SDK .csproj format since that is the new supported format and where all the fixes are going towards. Similarly, we would advice to move your project to target netcoreapp2.1 or above since 1.1 is no longer supported.

**Comment by RaghuMasineni** [comment]
@joperezr Thanks for the update.

Our Website was built using visual studio 2015 with .netcore 1.1.0. So the project file is still project.json .And there are no plans for using VS 2017 in near future.
Is it possible to migrate project.json to SDK.csProj and dotnet core 1.1 to 2.1 with visual studio 2015 ?
Please provide me any useful link if possible

Following are the error details in Jenkins for the issue reported,
-------------------------------------------------------------------
C:/*/*/*/*/refs/System.Net.Security.dll , CVE-2017-0249 , Severity: High 
An elevation of privilege vulnerability exists when the ASP.NET Core fails to properly sanitize web requests.
 
CVSS: 7.5
NVD URL: CVE-2017-0249 
Category: CWE-20 Improper Input Validation 
References: 
Source: MISC 
Name: https://github.com/aspnet/Announcements/issues/239 
URL: https://github.com/aspnet/Announcements/issues/239 
-----------------------------------------------------------------------
I have followed the steps mentioned in above link(adding direct dependency to 4.3.1) but still getting the issue

**Comment by RaghuMasineni** [comment]
Yes @Joperezr , I have tried with 4.3.2 by updating dotnet core to 1.1.4 fro 1.1.0.It did not fix the issue.
As of now I have proposed for upgrading current dotnet core 1.1  to core 2.* and waiting for confirmation.

But still I could not understand how using 4.3.1 or 4.3.2 for System.Net.Security is not fixing the issue as mentioned in https://github.com/dotnet/announcements/issues/12

**Comment by joperezr** [comment]
System.Net.Security does no longer ship as a separate package after .NET Core 2.0, and it is now part of the framework. That means that once you retarget to core 2.* then you won't have to have a package reference to System.Net.Security any longer. Although I'm not 100% sure, it is very possible that the fix for this problem you are seeing came AFTER shiping 4.3.2 package, which means that the only way to get the fix, would be to re-target to .NET Core 2.*

**Comment by joperezr** [comment]
@RaghuMasineni where you able to re-target your app to 2.*? Can this issue be closed now?

**Comment by RaghuMasineni** [comment]
@joperezr  Fix for the issue is available in 4.3.1 onwards as per below announcement,
https://github.com/dotnet/announcements/issues/12 

We are yet to migrate our application to Core 2.* due to other issues. You can close this thread but issue is not resolved and we are just ignoring these warnings for now.
Thanks.
